src/main/target/SITL/target.c

```diff
@@ -0,0 +1,411 @@
+
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <math.h>
+#define printf printf
+#define sprintf sprintf
+
+#include <errno.h>
+#include <time.h>
+#include <pthread.h>
+
+#include "drivers/io.h"
+#include "drivers/dma.h"
+#include "drivers/serial.h"
+#include "drivers/serial_tcp.h"
+#include "drivers/system.h"
+#include "drivers/pwm_output.h"
+#include "drivers/light_led.h"
+
+#include "drivers/timer.h"
+#include "drivers/timer_def.h"
+const timerHardware_t timerHardware[USABLE_TIMER_CHANNEL_COUNT] = {};
+
+#include "drivers/accgyro_fake.h"
+#include "flight/imu.h"
+
+#include "dyad.h"
+#include "target/SITL/udplink.h"
+
+static fdm_packet fdmPkt;
+static servo_packet pwmPkt;
+
+static struct timespec start_time;
+static double simRate = 1.0;
+static pthread_t tcpWorker, udpWorker;
+static bool workerRunning = true;
+static udpLink_t stateLink, pwmLink;
+static pthread_mutex_t updateLock;
+
+#define RAD2DEG (180.0 / M_PI)
+#define ACC_SCALE (256 / 9.81)
+#define GYRO_SCALE (8192 / 2000.0)
+void updateState(const fdm_packet* pkt) {
+	static double last_timestamp = 0; // in seconds
+	static uint64_t last_realtime = 0; // in uS
+
+	const uint64_t realtime_now = micros64_real();
+	if(realtime_now > last_realtime + 500*1e3) { // 500ms timeout
+		last_timestamp = 0;
+		last_realtime = 0;
+	}
+
+	const double deltaSim = pkt->timestamp - last_timestamp;  // in seconds
+	if(deltaSim < 0) { // don't use old packet
+		pthread_mutex_unlock(&updateLock);
+		return;
+	}
+
+	int16_t x,y,z;
+	x = -pkt->imu_linear_acceleration_xyz[0] * ACC_SCALE;
+	y = -pkt->imu_linear_acceleration_xyz[1] * ACC_SCALE;
+	z = -pkt->imu_linear_acceleration_xyz[2] * ACC_SCALE;
+	fakeAccSet(x, y, z);
+//	printf("[acc]%lf,%lf,%lf\n", pkt->imu_linear_acceleration_xyz[0], pkt->imu_linear_acceleration_xyz[1], pkt->imu_linear_acceleration_xyz[2]);
+
+	x = pkt->imu_angular_velocity_rpy[0] * GYRO_SCALE * RAD2DEG;
+	y = -pkt->imu_angular_velocity_rpy[1] * GYRO_SCALE * RAD2DEG;
+	z = -pkt->imu_angular_velocity_rpy[2] * GYRO_SCALE * RAD2DEG;
+	fakeGyroSet(x, y, z);
+//	printf("[gyr]%lf,%lf,%lf\n", pkt->imu_angular_velocity_rpy[0], pkt->imu_angular_velocity_rpy[1], pkt->imu_angular_velocity_rpy[2]);
+
+#if defined(SKIP_IMU_CALC)
+	double qw = pkt->imu_orientation_quat[0];
+	double qx = pkt->imu_orientation_quat[1];
+	double qy = pkt->imu_orientation_quat[2];
+	double qz = pkt->imu_orientation_quat[3];
+	double ysqr = qy * qy;
+	double xf, yf, zf;
+
+	// roll (x-axis rotation)
+	double t0 = +2.0 * (qw * qx + qy * qz);
+	double t1 = +1.0 - 2.0 * (qx * qx + ysqr);
+	xf = atan2(t0, t1) * RAD2DEG;
+
+	// pitch (y-axis rotation)
+	double t2 = +2.0 * (qw * qy - qz * qx);
+	t2 = t2 > 1.0 ? 1.0 : t2;
+	t2 = t2 < -1.0 ? -1.0 : t2;
+	yf = asin(t2) * RAD2DEG; // from wiki
+
+	// yaw (z-axis rotation)
+	double t3 = +2.0 * (qw * qz + qx * qy);
+	double t4 = +1.0 - 2.0 * (ysqr + qz * qz);
+	zf = atan2(t3, t4) * RAD2DEG;
+	imuSetAttitudeRPY(xf, -yf, zf); // yes! pitch was inverted!!
```

Comment:
Isn't it possible to set quaternion and let existing code compute euler angles?
Also some parts of rotation matrix may be used in code (maybe angle from level for throttle compensation; arm threshold)